File: tests/governance.test.ts
import { MultiSigContract } from "../governance/contract";

describe("MultiSig Governance Contract", () => {
  const council = ["val1", "val2", "val3", "val4", "val5"];
  const contract = new MultiSigContract(council, 3); // 3-of-5 quorum

  test("should reach quorum with 3 signatures", () => {
    contract.sign("val1");
    contract.sign("val2");
    contract.sign("val3");

    expect(contract.hasQuorum()).toBe(true);
    expect(contract.executeProposal("Allocate Community Fund")).toBe(true);
  });

  test("should fail quorum with only 2 signatures", () => {
    const contract2 = new MultiSigContract(council, 3);
    contract2.sign("val1");
    contract2.sign("val2");

    expect(contract2.hasQuorum()).toBe(false);
    expect(contract2.executeProposal("Fail Proposal")).toBe(false);
  });
});

File: tests/governance_test.go
package tests

import (
    "testing"
    "stg-chain/governance"
)

// Mock council of 5 validators
var council = []string{"val1", "val2", "val3", "val4", "val5"}

func TestMultiSigQuorum(t *testing.T) {
    contract := governance.NewMultiSigContract(council, 3) // 3-of-5 quorum

    // Simulate signatures
    contract.Sign("val1")
    contract.Sign("val2")
    contract.Sign("val3")

    if !contract.HasQuorum() {
        t.Errorf("Expected quorum reached with 3 signatures, got false")
    }

    // Execute proposal
    executed := contract.ExecuteProposal("Allocate Community Fund")
    if !executed {
        t.Errorf("Expected proposal execution, got false")
    }
}

func TestInsufficientQuorum(t *testing.T) {
    contract := governance.NewMultiSigContract(council, 3)

    contract.Sign("val1")
    contract.Sign("val2")

    if contract.HasQuorum() {
        t.Errorf("Quorum should not be reached with only 2 signatures")
    }
}
